pattern/constraint/thing/ValueConstraint.java

```diff
@@ -326,6 +363,22 @@ public java.lang.String toString() {
         public String clone(Conjunction.ConstraintCloner cloner) {
             return cloner.cloneVariable(owner).valueString(predicate(), value);
         }
+
+        @Override
+        public boolean isCompatibleWithEQ(ValueConstraint<?> conclusionValueConstraint) {
+            if (conclusionValueConstraint.isVariable()) return true;
+            if (!conclusionValueConstraint.isString()) return false;
+            if (predicate.isEquality()) {
+                return PredicateOperator.Equality.of(predicate.asEquality()).apply(Predicate.compareStrings(conclusionValueConstraint.asString().value(), value));
+            } else if (predicate.isSubString()) {
+                PredicateOperator.SubString operator = PredicateOperator.SubString.of(predicate.asSubString());
+                if (operator == PredicateOperator.SubString.CONTAINS) {
+                    return PredicateOperator.SubString.CONTAINS.apply(conclusionValueConstraint.asString().value(), value);
```

Comment:
LIKE and CONTAINS actually have two different predicate value types - Pattern and String.
(This) ValueConstraint class has only the String. 

Introducing a type-parameter on PredicateOperator.SubString is a workaround.